http-api/src/v1/projects.rs

```diff
@@ -205,50 +222,53 @@ async fn history_handler(
         verified,
     } = qs;
 
+    let projects = ctx.projects.read().await;
+
     let (sha, fallback_to_head) = match parent {
-        Some(commit) => (commit, false),
+        Some(commit) => {
+            let oid = git::Oid::from_str(&commit)?;
+            (oid, false)
+        }
         None => {
             let info = ctx.project_info(project.to_owned()).await?;
 
             if let Some(head) = info.head {
-                (head.to_string(), true)
+                (head, true)
             } else {
                 return Err(Error::NoHead("project head is not set"));
             }
         }
     };
 
-    let reference = Reference::head(
-        Namespace::from(project.to_owned()),
-        None,
-        One::from_str(&sha).map_err(|_| Error::NotFound)?,
-    );
-
-    let commits = browse(reference, ctx.paths.to_owned(), |browser| {
-        radicle_source::commits::<PeerId>(browser, None)
-    })
-    .await?;
+    let repo = git::Repository::new(ctx.paths.git_dir()).map_err(Error::from)?;
+    let browser = git::Browser::new_with_namespace(
+        &repo,
+        &git::Namespace::try_from(project.encode_id().as_str())
+            .map_err(|_| Error::MissingNamespace)?,
+        sha,
+    )
+    .map_err(Error::from)?;
+    let commits = browser.get();
+    let stats = browser.get_stats()?;
```

Comment:
The problem with using surf here is that it always fetches the entire commit history, so this will be really slow for large repos.

Reply:
As talked over we keep fetching for now the entire commit history, but since we are paginating we at least do not return all the commits.